From 7b3293e449c2fbc4a79bad0321db76fb647a755f Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:16:30 -0400
Subject: [net] netfilter: nf_tables: accept QUEUE/DROP verdict parameters

Message-id: <20141021161629.22888.96709.stgit@localhost.localdomain>
Patchwork-id: 97916
O-Subject: [RHEL7 net PATCH 019/124] netfilter: nf_tables: accept QUEUE/DROP verdict parameters
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: e0abdadcc6e113ed2e22c85b350074487095875b

 commit e0abdadcc6e113ed2e22c85b350074487095875b
 Author: Patrick McHardy <kaber@trash.net>
 Date:   Tue Feb 18 18:06:50 2014 +0000

    netfilter: nf_tables: accept QUEUE/DROP verdict parameters

    Allow userspace to specify the queue number or the errno code for QUEUE
    and DROP verdicts.

    Signed-off-by: Patrick McHardy <kaber@trash.net>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 4b7e14d..0b56340 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -3174,9 +3174,16 @@ static int nft_verdict_init(const struct nft_ctx *ctx, struct nft_data *data,
  data->verdict = ntohl(nla_get_be32(tb[NFTA_VERDICT_CODE]));
 
  switch (data->verdict) {
- case NF_ACCEPT:
- case NF_DROP:
- case NF_QUEUE:
+ default:
+  switch (data->verdict & NF_VERDICT_MASK) {
+  case NF_ACCEPT:
+  case NF_DROP:
+  case NF_QUEUE:
+   break;
+  default:
+   return -EINVAL;
+  }
+  /* fall through */
  case NFT_CONTINUE:
  case NFT_BREAK:
  case NFT_RETURN:
@@ -3197,8 +3204,6 @@ static int nft_verdict_init(const struct nft_ctx *ctx, struct nft_data *data,
   data->chain = chain;
   desc->len = sizeof(data);
   break;
- default:
-  return -EINVAL;
  }
 
  desc->type = NFT_DATA_VERDICT;
-- 
1.7.1